ethereum/contracts/query/QueryResponse.sol

```diff
@@ -0,0 +1,246 @@
+
+        uint8 respNumResults;
+        (respNumResults, respIdx) = pcr.response.asUint8Unchecked(respIdx);
+        if (respNumResults != numBatchCallData) {
+                revert UnexpectedNumberOfResults();
+        }
+
+        r.result = new EthCallData[](numBatchCallData);
+
+        // Walk through the call data and results in lock step.
+        for (uint idx = 0; idx < numBatchCallData;) {
+            (r.result[idx].contractAddress, reqIdx) = pcr.request.asAddressUnchecked(reqIdx);
+
+            (len, reqIdx) = pcr.request.asUint32Unchecked(reqIdx); // call_data_len
+            (r.result[idx].callData, reqIdx) = pcr.request.sliceUnchecked(reqIdx, len);
+
+            (len, respIdx) = pcr.response.asUint32Unchecked(respIdx); // result_len
+            (r.result[idx].result, respIdx) = pcr.response.sliceUnchecked(respIdx, len);
+
+            unchecked { ++idx; }
+        }
+
+        checkLength(pcr.request, reqIdx);
+        checkLength(pcr.response, respIdx);
+        return r;
+    }
+
+    /**
+     * @dev verifyQueryResponseSignatures verifies the signatures on a query response. It calls into the Wormhole contract.
+     * IWormhole.Signature expects the last byte to be bumped by 27 
+     * see https://github.com/wormhole-foundation/wormhole/blob/637b1ee657de7de05f783cbb2078dd7d8bfda4d0/ethereum/contracts/Messages.sol#L174
+     */
+    function verifyQueryResponseSignatures(address _wormhole, bytes memory response, IWormhole.Signature[] memory signatures) public view {
+        IWormhole wormhole = IWormhole(_wormhole);
+        // TODO: make a verifyCurrentQuorum call on the core bridge so that there is only 1 cross call instead of 4
+        uint32 gsi = wormhole.getCurrentGuardianSetIndex();
+        IWormhole.GuardianSet memory guardianSet = wormhole.getGuardianSet(gsi);
+
+        bytes32 responseHash = getResponseDigest(response);
+
+       /**
+        * @dev Checks whether the guardianSet has zero keys
+        * WARNING: This keys check is critical to ensure the guardianSet has keys present AND to ensure
+        * that guardianSet key size doesn't fall to zero and negatively impact quorum assessment.  If guardianSet
+        * key length is 0 and vm.signatures length is 0, this could compromise the integrity of both vm and
+        * signature verification.
+        */
+        if(guardianSet.keys.length == 0){
+            revert("invalid guardian set");
+        }
+
+       /**
+        * @dev We're using a fixed point number transformation with 1 decimal to deal with rounding.
+        *   WARNING: This quorum check is critical to assessing whether we have enough Guardian signatures to validate a VM
+        *   if making any changes to this, obtain additional peer review. If guardianSet key length is 0 and
+        *   vm.signatures length is 0, this could compromise the integrity of both vm and signature verification.
+        */
+        if (signatures.length < wormhole.quorum(guardianSet.keys.length)){
```

Comment:
Is there a reason we don't implement the quorom function in this library? That would reduce the number of cross calls to the Wormhole contract.

Reply:
This may be considered in a follow up PR.